engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java

```diff
@@ -193,19 +199,35 @@ public Rope getLiteralSource() {
    * @param source the module source.
    */
   public void setLiteralSource(String source) {
-    setLiteralSource(Rope.apply(source));
+    setLiteralSource(Rope.apply(source), null, null);
   }
 
   /**
    * Sets new literal sources for the module.
    *
    * @param source the module source.
    */
-  public void setLiteralSource(Rope source) {
-    this.literalSource = source;
-    this.compilationStage = CompilationStage.INITIAL;
-    this.cachedSource = null;
+  public void setLiteralSource(Rope source, IR$Literal$Number change, model.TextEdit edit) {
     this.isInteractive = true;
+    if (this.scope != null && edit != null) {
+      if (this.scope.simpleUpdate(edit)) {
+        this.sources = this.sources.newWith(source);
+        final Function1<IR.Expression, IR.Expression> fn = new Function1<IR.Expression, IR.Expression>() {
```

Comment:
Create a wrapper for `mapExpressions` in Scala that will take a java.util.Function as an argument and use a lambda here – it's much easier than going this way.

Reply:
There is not a functional difference  between `Function1` and `java.util.Function`. They both work the same. Only, if we want to _keep engine clean_ from Scala concepts, we should prefer `java.util.Function`.

> use a lambda here

I could use lambda even with `Function1`. However I need a reference to the function itself, so I can recursively apply it: `return v1.mapExpressions(this);` - that's the reason why I decided to define the function as an anonymous inner class.

However there is yet another problem with this `mapExpressions` - I am producing copy of the whole `IR`. Even the parts that weren't changed (as `mapExpressions` implementation always use `copy` of appropriate case class). Ideally we just replace the path from root to the affected literal...